### ISSUE TITLE
Lost/hanging requests with 3.6.2

### ISSUE THREAD

**Issue by pvizeli** [aio-libs/aiohttp#4258]
## Long story short

Based on: https://github.com/home-assistant/hassio/pull/1342

We run into an issue that requests produce timeouts and were never logged by our middleware they protocol/monitor requests.  All instances with this issue running with an nginx proxy manager in front. Do we modify the URL with a parameter, it worked. It is very random. The only way to fix this issue was to downgrade aiohttp.

We can see the request is going to instance but we can't see it on the application layer. Maybe interesting ist that all the requests use cookies for session handling. 

## Expected behaviour

Everything work like 3.6.1

## Actual behaviour

Random connections run into a timeouts.

## Steps to reproduce

Run an nginx proxymanager in front of aiohttp 3.6.2 and use a site with cookies.

## Your environment

Python 3.7 with uvloop and without. We use aioHTTP as server application in this case.

**Comment by webknjaz** [comment]
> without uvloop doesn't have this issue

did you report this to uvloop?

**Comment by pvizeli** [comment]
EDIT: It look like without uvloop it doesn't have this issue.

I would do this after someone can point me to the part of code they changed on 3.6.2 and can affect this. Maybe @1st1 can assist me later.

EDIT: we had now the issue without uvloop @webknjaz. It is very random. So with 30 requests, 1-2 requests hangs.

**Comment by webknjaz** [comment]
@pvizeli any chance you could `git bisect` this? Also, please post the log of `pip install`, we need to know if you use C-extensions or not and if you're installing from wheels or building stuff during install time.

**Comment by webknjaz** [comment]
Source comparison: https://github.com/aio-libs/aiohttp/compare/v3.6.1...v3.6.2#files_bucket

**Comment by webknjaz** [comment]
Changelog: https://docs.aiohttp.org/en/stable/changes.html#id1

**Comment by webknjaz** [comment]
Looking at https://pypi.org/project/aiohttp/3.6.1/#files vs. https://pypi.org/project/aiohttp/3.6.2/#files: you might've hit a wheel with `py3-none-any` marker which would mean that you're running a pure-python version so `pip install` is really important here. 
Until you post this info, we have no idea what you're really running and consequently how to reproduce it.

**Comment by pvizeli** [comment]
We build our own wheel (on Alpine) and we use the C extension. I can get a freeze from that docker if you want? The affected container is `homeassistant/amd64-hassio-supervisor:190`

**Comment by pvizeli** [comment]
Not sure, but: https://github.com/mjpieters/aiohttp/blob/96e441a435a216a15e0b64ff140d01eb075761ec/aiohttp/client_proto.py#L179-L189

They now init a timeout handler on receive data, also if data are empty. I think that needs to be after None check of data.

**Comment by webknjaz** [comment]
Have you seen this? https://twitter.com/1st1/status/1187214384625541121?s=19

**Comment by webknjaz** [comment]
cc @mjpieters ^

**Comment by webknjaz** [comment]
@pvizeli if you could come up with a PR that works for you, that would be super helpful. This also needs regression tests so that it won't happen again.

**Comment by mjpieters** [comment]
So the timeout would need to be reset only if `data` is not empty? That sounds more than sensible to me, would be happy to make that  change.

@asvetlov can you confirm this is the case?

**Comment by mjpieters** [comment]
Actually, there is **no `sock_read` timeout set** by default, only an overall timeout. My changes did not alter this, they only affect client connections where you explicitly used a `aiohttp.ClientTimeout(sock_read=....)` object as the `timeout` parameter for your requests. I don't see any such use in the home-assistant source code; *all* client HTTP requests either use `timeout=integer_value` or `timeout=None`.

In short, I'm a little sceptical you have located the correct issue here.

What happens now, for home-assistant:

* If you don't use a timeout, you get a `ClientTimeout(total=5*60)` object, which leaves `sock_read` set to `None`, the default. If you use a numeric value or `None`, then a `ClientTimeout()` object is created with `total` set to the given value. `total=None` just means there is no overall timeout.

* `sock_read` is passed to the `ResponseHandler.set_response_params()` method, and stored as `self._read_timeout`. There is also a `self._read_timeout_handle` attribute, initially set to `None`
* When `data_received()` calls `self._reschedule_timeout()` and `read_timeout` is *not set* then *nothing happens*:

    ```python
    def _reschedule_timeout(self) -> None:
        timeout = self._read_timeout
        if self._read_timeout_handle is not None:
            self._read_timeout_handle.cancel()

        if timeout:
            self._read_timeout_handle = self._loop.call_later(
                timeout, self._on_read_timeout)
        else:
            self._read_timeout_handle = None
    ```

The issue is elsewhere, I'm afraid. You say that there are timeouts? Where are those timeouts? Exactly what Python exception, if any, is being raised? What is the traceback attached to those timeouts?

**Comment by pvizeli** [comment]
@webknjaz we don't use Python 3.8 currently. Because we use a lot of python packages and that take time until all are updated :)

@mjpieters It was only a feeling. I try next week to debug the issue a bit deeper. It's only happening with requests from/over Nginx. Not sure what they do differently.

**Comment by asvetlov** [comment]
1. If it is a timeout -- an exception & stack trace can help.
2. Rescheduling sock_timeout checker even if no data doesn't harm it can *reduce* a chance of timeout.

Do you have a scenario to reproduce the issue locally? Preferable a small script. Without additional info, I have no idea what's going wrong. No commit between v3.6.1 and v3.6.2 tags looks suspicious.

**Comment by pvizeli** [comment]
However we had issues with Home Assistant and Hass.io if one of them use 3.6.2 with nginx proxy-manager in front of it.

I try to get more details out of aiohttp core

**Comment by asvetlov** [comment]
@webknjaz suggested bisecting aiohttp to find what commit introduced the problem.
The diff is about 15 commits; the check should not take too long probably.

**Comment by frenck** [comment]
After tons of testing with Home Assistant/Hass.io on aiohttp 3.6.2 and Nginx (1.6) as a reverse proxy, I've been able to boil it down to the following:

Not working NGinx proxy configuration on aiohttp 3.6.2 (working on 3.6.1):

```txt
http {
  ...

  server {
    ...

    location / {
      proxy_pass http://backend;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
    }
  }
}
```

Working NGinx proxy configuration on both aiohttp 3.6.2 and 3.6.1

```txt
http {
  ...

  map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
  }

  server {
    ...

    location / {
      proxy_pass http://backend;
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
    }
  }
}
```

For clarity; the difference is in the mapping of the `Connection` header.

In the first example, the connection will hang until the timeout kicks in after 60 seconds.

This probably means, with keep-alive, that on explicit request of closing the connection from the client, the issue disappears, however, this also means the server will not initiate to send a close header over the connection when it is done? (If that makes sense in any way in this case).

I hope this helps with finding the issue.

**Comment by pvizeli** [comment]
@webknjaz @asvetlov maybe I was not complete wrong with https://github.com/aio-libs/aiohttp/issues/4258#issuecomment-545510231

**Comment by pvizeli** [comment]
Short Update, maybe we used the python extension and not the cython extension. Since 3.6.2 push a wheel for pure python to PyPI, is pretty difficult to don't use that one and instead build a cython version. It would be nice, we the pure python wheels are not pushed to pypi.

We make a new test from our https://wheels.home-assistant.io/ which have now the cython version available for musl+jemalloc

**Comment by Dreamsorcerer** [comment]
This could have been caused by one of the parser bugs that has been fixed more recently. So, if someone could potentially retest (@bdraco not sure if you're familiar with this part of homeassistant?), that'd be great.

**Comment by bdraco** [comment]
I have a feeling this is likely been fixed. Let me ask around.

**Comment by bdraco** [comment]
It looks like the issue is no longer reproducible https://github.com/home-assistant/core/pull/38082

**Comment by Dreamsorcerer** [comment]
From what I see above, the issue was only with the Python parser, which was used by mistake. So, I assume the compiled wheels were used in order to successfully upgrade to 3.6.2, and then nobody tried it since..

**Comment by bdraco** [comment]
Oh, that makes sense. HA has since replaced the wheel building method so its very likely that was what happened and why HA can't repro it anymore.  Its possible possible it got solve in when of the Upgrade fixes in 3.9.x. Should we leave it open?

**Comment by Dreamsorcerer** [comment]
It's probably fixed, so if no-one is going to retest, then we'll just leave it closed.
